File: GenericProducerConsumer/src/Producer.java
import java.util.Queue;

public class Producer <MyType>{
    Queue<MyType> queue;
    public Producer(Queue<MyType> queue) {
        this.queue = queue;
    }
    public void produce(MyType item) {
        queue.add(item);
        System.out.println("Adding item to queue: " + item);
    }
}

File: os/producer_consumer/producer.cpp
#include <string>
#include <queue>
using namespace std;

template <typename T>
class producer {
    private:
        queue<T>* q;
    public:
        producer(queue<T>* q) {
            this->q = q;
        }
        void produce(T item) {
            q->push(item);
        }
};